src/HttpServer.php

```diff
@@ -60,50 +64,115 @@ public function addMiddleware($middleware)
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
-                    }
+    /** @internal */
+    public function handleBody(RequestInterface $request, ConnectionInterface $connection)
+    {
+        $input = $this->sendHttpBodyStream($request, $connection);
+
+        $connection->removeAllListeners('data');
+        $that = $this;
+
+        if ($this->isChunkedEncodingActive($request)) {
+            $connection->on('data', function ($data) use ($input){
+                $input->emit('data', array($data));
+            });
+            return;
+        }
+
+        if (!$request->hasHeader('Content-Length')) {
+            $connection->on('data', function ($data) use ($that, $input, $connection){
+                if ($data === '') {
+                    // Simple request without body and without 'Content-Length'
+                    $input->emit('end', array());
+                    return;
                 }
+                // Request hasn't defined 'Content-Length' and body given
+                $that->sendResponse(new Response(411), $connection);
+            });
+            return;
+        }
+
+        $contentLength = $request->getHeaderLine('Content-Length');
+
+        $int = (int) $contentLength;
+        if ((string)$int !== (string)$contentLength) {
+            // Send 400 status code if the value of 'Content-Length'
+            // is not an integer or is duplicated
+            $this->sendResponse(new Response(400), $connection);
+            return;
+        }
+
+        $transferredLength = 0;
+
+        $connection->on('data', function ($data) use ($that, $contentLength, &$transferredLength, $input) {
+            // 'Content-Length' is given
+            if (($transferredLength + strlen($data)) > $contentLength) {
+                // Only emit data until the value of 'Content-Length' is reached, the rest will be ignored
+                $data = substr($data, 0, $contentLength - $transferredLength);
+            }
+
+            $transferredLength += strlen($data);
+            $input->emit('data', array($data));
+
+            if ((int)$transferredLength === (int)$contentLength) {
+                // 'Content-Length' reached, stream will end
+                $input->emit('end', array());
             }
         });
     }
 
+    /** @internal */
+    public function sendHttpBodyStream(RequestInterface $request, ConnectionInterface $connection)
+    {
+        $input = new ReadableStream();
```

Comment:
Note that ReadableStream does not support backpressure and I'm currently looking into removing this for the next stream release.